dali/test/python/operator/test_split_merge.py

```diff
@@ -0,0 +1,146 @@
+        out, = pipe_sm.run()
+        out_true, = pipe_true.run() if data_true else ([], )
+        out_false, = pipe_false.run() if data_false else ([], )
+        out_baseline = []
+        idx_true = 0
+        idx_false = 0
+        for p in predicate:
+            if p:
+                out_baseline.append(out_true[idx_true])
+                idx_true = idx_true + 1
+            else:
+                out_baseline.append(out_false[idx_false])
+                idx_false = idx_false + 1
+        if dev == "gpu":
+            out = [out[i].as_cpu() for i in range(bs)]
+            out_baseline = [out_baseline[i].as_cpu() for i in range(bs)]
+        check_batch(out, out_baseline, bs)
+
+
+def test_conditional_split_merge():
+    rng = np.random.default_rng()
+    for dev in ["cpu", "gpu"]:
+        for pred_gen in [
+                lambda x: np.array(x < 3), lambda x: np.array(x % 2 == 0),
+                lambda x: np.array(x % 3 == 0), lambda _: np.array(False),
+                lambda _: rng.choice([np.array(True), np.array(False)])
+        ]:
+            yield check_conditional_split_merge, dev, pred_gen
+
+
+@pipeline_def
+def conditional_split_merge_reinterpret_pipe(dtype, layout, shape):
+    batch_size = Pipeline.current().max_batch_size
+    input = fn.external_source(
+        source=[[np.full((10, 10, 3), 42, dtype=np.int32) for _ in range(batch_size)]], cycle=True)
+    pred = fn.external_source(
+        source=[[np.array(i % 2 == 0, dtype=np.bool) for i in range(batch_size)]], cycle=True)
+    true_branch, false_branch = fn._conditional.split(input, predicate=pred)
+    false_changed = fn.reinterpret(false_branch, dtype=dtype, layout=layout, shape=shape)
+    return fn._conditional.merge(true_branch, false_changed, predicate=pred)
+
+
+def run_conditional_split_merge_reinterpret(dtype, layout, shape):
+    bs = 10
+    kwargs = {
+        "batch_size": bs,
+        "num_threads": 4,
+        "device_id": 0,
+        "prefetch_queue_depth": 1  # so that it's easier to use external source
+    }
+    pipe = conditional_split_merge_reinterpret_pipe(dtype, layout, shape, **kwargs)
+    pipe.build()
+    pipe.run()
+
+
+def check_fail_conditional_split_merge(dtype, layout, shape, err_glob):
+    with assert_raises(RuntimeError, glob=err_glob):
+        run_conditional_split_merge_reinterpret(dtype, layout, shape)
+
+
+def test_fail_conditional_split_merge():
+    base = (
+        "Divergent data found in different branches of conditional operation. All paths in "
+        "conditional operation are merged into one batch which must have consistent type, "
+        "dimensionality, layout and other metadata. Found distinct "
+    )
+    yield check_fail_conditional_split_merge, types.UINT32, None, None, base + "types*"
+    yield check_fail_conditional_split_merge, None, "HWC", None, base + "layouts*"
+    yield check_fail_conditional_split_merge, None, None, [10, -1], base + "sample dimensions*"
```

Comment:
Small hint. Now that we have `nose2`, you can write the tests without abusing the generators. I believe implementation using `@param` would be way better. Like here: https://github.com/NVIDIA/DALI/pull/4379/files?show-viewed-files=true&file-filters%5B%5D=#diff-9d4f5c833015a5d47d5c41f94b79bd93651bc9e75c3e8debb52aab2d6faa97c9R119

I've used it with `unittest.TestCase` class, but I guess it works with `test_` functions as well.

Reply:
I'm waiting on the promised guideline on how to write the tests for the new framework. 
Your example only shows one parameter though, so I'm gonna stick with what I have for now.

Reply:
Done*